contracts/DecentHats.sol

```diff
@@ -0,0 +1,393 @@
+import {ISablierV2LockupLinear} from "./interfaces/sablier/ISablierV2LockupLinear.sol";
+
+contract DecentHats {
+    string public constant NAME = "DecentHats";
+    bytes32 public constant SALT =
+        0x5d0e6ce4fd951366cc55da93f6e79d8b81483109d79676a04bcc2bed6a4b5072;
+
+    struct SablierStreamParams {
+        ISablierV2LockupLinear sablier;
+        address sender;
+        address asset;
+        LockupLinear.Timestamps timestamps;
+        Broker broker;
+        uint128 totalAmount;
+        bool cancelable;
+        bool transferable;
+    }
+
+    struct TermedParams {
+        uint128 termEndDateTs;
+        address[] nominatedWearers;
+    }
+
+    struct Hat {
+        address wearer;
+        string details;
+        string imageURI;
+        SablierStreamParams[] sablierParams;
+        TermedParams[] termedParams;
+        uint32 maxSupply;
+        bool isMutable;
+        bool isTermed;
+    }
+
+    struct CreateTreeParams {
+        IHats hatsProtocol;
+        IERC6551Registry registry;
+        IHatsModuleFactory hatsModuleFactory;
+        ModuleProxyFactory moduleProxyFactory;
+        address decentAutonomousAdminMasterCopy;
+        address hatsAccountImplementation;
+        address keyValuePairs;
+        address hatsElectionEligibilityImplementation;
+        Hat adminHat;
+        Hat[] hats;
+        string topHatDetails;
+        string topHatImageURI;
+    }
+
+    /* /////////////////////////////////////////////////////////////////////////////
+                        EXTERNAL FUNCTIONS
+    ///////////////////////////////////////////////////////////////////////////// */
+    function createAndDeclareTree(CreateTreeParams calldata params) public {
+        (uint256 topHatId, address topHatAccount) = _createTopHatAndAccount(
+            params.hatsProtocol,
+            params.topHatDetails,
+            params.topHatImageURI,
+            params.registry,
+            params.hatsAccountImplementation
+        );
+
+        _updateKeyValuePairs(params.keyValuePairs, topHatId);
+
+        (uint256 adminHatId, ) = _createAdminHatAndAccount(
+            params.hatsProtocol,
+            params.registry,
+            params.moduleProxyFactory,
+            params.decentAutonomousAdminMasterCopy,
+            params.hatsAccountImplementation,
+            topHatAccount,
+            topHatId,
+            params.adminHat
+        );
+
+        for (uint256 i = 0; i < params.hats.length; ) {
+            if (params.hats[i].isTermed) {
+                // Create election module and set as eligiblity
+                _createTermedHatAndAccountAndMintAndStreams(
+                    params.hatsProtocol,
+                    topHatAccount,
+                    _createElectionEligiblityModule(
+                        params.hatsModuleFactory,
+                        params.hatsElectionEligibilityImplementation,
+                        params.hatsProtocol.getNextId(adminHatId),
```

Comment:
Has this been tested when there are multiple termed hats being created?

Reply:
but still, definitely need some tests written

Reply:
got the tree test. Need to add a test for `createTermedRoleHat` external function

Reply:
addressed